### PR TITLE
More method for db_insert returning id

### PR DESCRIPTION
**Pull request by albertlast**
Today i notice that i got a wrong understanding from replace method of mysql,
so i was able to add replace to supported method for returning the ai.
So only insert ingore keep unsupported,
i would try later this day to add also for ingore.

### CODE REVIEW

**Comment by albertlast** [comment]
So this pr is done,
would be good if someone is able to test specialy the ingore part,
is very complex the solution...

File: Sources/Subs-Db-mysql.php [hunk 1]
```diff
@@ -776,13 +776,24 @@ function smf_db_insert($method = 'replace', $table, $columns, $data, $keys, $ret
 	global $smcFunc, $db_connection, $db_prefix;
 
 	$connection = $connection === null ? $db_connection : $connection;
+	
+	$return_var;
```
Comment (jdarwood007) [inline Sources/Subs-Db-mysql.php]: Typo?

File: Sources/Subs-Db-mysql.php [hunk 2]
```diff
@@ -811,24 +822,82 @@ function smf_db_insert($method = 'replace', $table, $columns, $data, $keys, $ret
...
+					$where_string += key($indexed_columns[$x]) . ' = '. $insertRows[$i][$x];
+					if (($x + 1) < $count2)
+						$where_string += ' AND ';
+				}
+
+				$request = $smcFunc['db_query']('','
+					SELECT `'. $keys[0] . '` FROM ' . $table .'
+					WHERE ' . $where_string . ' LIMIT 1',
+					array()
+				);
```
Comment (jdarwood007) [inline Sources/Subs-Db-mysql.php:879]: Only issue I have with this, is we are assuming that $keys[0] contains the PK.  I should be true for 99% of cases, but I wonder if that is still true.

I believe the entire need for us specifying the PK in the builder was SQLite had issues with inserts and had to rebuild indexes or something of the sort.  I don't think its even relevant for just MySQL and PostgreSQL.

File: Sources/Subs-Db-mysql.php (hunk 1 above)
Comment (albertlast) [reply Sources/Subs-Db-mysql.php]: i need the var across the methode,
to get the right scrope i defined this var on start.

File: Sources/Subs-Db-mysql.php (hunk 2 above)
Comment (albertlast) [reply Sources/Subs-Db-mysql.php:879]: The pk field is essential for pg and
mysql only allow that one field in a table is a ai field,
so i had to assume that the first pk field is the ai field.

similiar logic you find on pg side: https://github.com/SimpleMachines/SMF2.1/blob/release-2.1/Sources/Subs-Db-postgresql.php#L769
like i said this is for the limitation on mysql side.

File: Sources/Subs-Db-mysql.php (hunk 1 above)
Comment (jdarwood007) [reply Sources/Subs-Db-mysql.php]: Shouldn't you specify a = null then?  Just seems weird to see it like this.

File: Sources/Subs-Db-mysql.php (hunk 2 above)
Comment (jdarwood007) [reply Sources/Subs-Db-mysql.php:879]: Well developers should be following the proper logic and making the AI field the first in the list for the keys.  Doesn't mean it will happen 100%, but it could.  I guess we call those bugs in that code.

File: Sources/Subs-Db-mysql.php (hunk 1 above)
Comment (albertlast) [reply Sources/Subs-Db-mysql.php]: i hope empty() keeps triggered

Comment (jdarwood007) [reply Sources/Subs-Db-mysql.php]: https://secure.php.net/empty

Yes, null is considered empty.  But you can also test for null with is_null or === null.  Which would make it different than 0 or false.

File: Sources/Subs-Db-mysql.php (hunk 2 above)
Comment (albertlast) [reply Sources/Subs-Db-mysql.php:879]: so it's fine how it is?

Comment (Sesquipedalian) [reply Sources/Subs-Db-mysql.php:879]: Approval, @jdarwood007?

Comment (jdarwood007) [reply Sources/Subs-Db-mysql.php:879]: I've voiced my concerns, if no others agree, then merge it.  Worse case is it becomes a coding standard for the keys to contain the pki as the first item.  We could already use the passed pki info to figure it out ourselves.

Comment (Sesquipedalian) [reply Sources/Subs-Db-mysql.php:879]: I don't have an opinion either way. SQL is not my strong suit, so I'm just waiting to see if you two can reach a conclusion or if another dev decides to weigh in.
